clang/lib/Sema/SemaOpenACC.cpp

```diff
@@ -423,6 +450,52 @@ ExprResult SemaOpenACC::ActOnIntExpr(OpenACCDirectiveKind DK,
   return IntExpr;
 }
 
+ExprResult SemaOpenACC::ActOnVar(Expr *VarExpr) {
+  // We still need to retain the array subscript/subarray exprs, so work on a
+  // copy.
+  Expr *CurVarExpr = VarExpr->IgnoreParenImpCasts();
+
+  // Sub-arrays/subscript-exprs are fine as long as the base is a
+  // VarExpr/MemberExpr. So strip all of those off.
+  while (isa<ArraySectionExpr, ArraySubscriptExpr>(CurVarExpr)) {
+    if (auto *SubScrpt = dyn_cast<ArraySubscriptExpr>(CurVarExpr))
+      CurVarExpr = SubScrpt->getBase()->IgnoreParenImpCasts();
+    else
+      CurVarExpr =
+          cast<ArraySectionExpr>(CurVarExpr)->getBase()->IgnoreParenImpCasts();
+  }
```

Comment:
No, a cast is an 'expression', so that makes it not a 'var' reference by my interpretation of the spec.  So it is disallowed.

Reply:
My reading of the standard says 'no', it shouldn't?  I wouldn't consider that to be an 'array-element'.

Both of the existing implementations I have access to also reject it: https://godbolt.org/z/Pj98d6c5M

Reply:
> My reading of the standard says 'no', it shouldn't? I wouldn't consider that to be an 'array-element'.

I don't know, just asking, since it is very similar to what have in OpenMP :)

Reply:
>I don't know, just asking, since it is very similar to what we have in OpenMP :)

Yeah, so far I've realized that the OpenACC spec is a very under-specified version of OpenMP in a lot of situations :)  I figure I can always loosen these restrictions when/if someone complains.